contracts/preset/ERC721PsiMysteryBox.sol

```diff
@@ -0,0 +1,142 @@
+// SPDX-License-Identifier: MIT
+/**
+  ______ _____   _____ ______ ___  __ _  _  _ 
+ |  ____|  __ \ / ____|____  |__ \/_ | || || |
+ | |__  | |__) | |        / /   ) || | \| |/ |
+ |  __| |  _  /| |       / /   / / | |\_   _/ 
+ | |____| | \ \| |____  / /   / /_ | |  | |   
+ |______|_|  \_\\_____|/_/   |____||_|  |_|   
+                                              
+                                            
+ */
+
+pragma solidity ^0.8.0;
+
+import "@chainlink/contracts/src/v0.8/interfaces/VRFCoordinatorV2Interface.sol";
+import "@chainlink/contracts/src/v0.8/VRFConsumerBaseV2.sol";
+import "@openzeppelin/contracts/utils/Strings.sol";
+
+import "fpe-map/contracts/FPEMap.sol";
+
+import "../ERC721Psi.sol";
+
+abstract contract ERC721PsiMysteryBox is VRFConsumerBaseV2, ERC721Psi {
+    using Strings for uint256;
+
+    uint32 private constant _callbackGasLimit = 200000;
+
+    uint256 private seed;
+
+    event RandomnessRequest(uint256 requestId);
+
+
+    constructor() {
+        VRFConsumerBaseV2(_coordinator());
+    }
+
+    /**
+     * @dev The metadata URI before revealing.
+     */
+    function _unrevealedURI() internal pure virtual returns (string memory) {
+        return "";
+    }
+
+    /**
+     * @dev See {IERC721Metadata-tokenURI}.
+     */
+    function tokenURI(uint256 tokenId) public view virtual override returns (string memory) {
+        require(_exists(tokenId), "ERC721Psi: URI query for nonexistent token");
+
+        if(seed == 0) {
+            return _unrevealedURI();
+        }
+
+        string memory baseURI = _baseURI();
+
+        uint256 metadataId = FPEMap.fpeMappingFeistelAuto(tokenId, seed, _maxSupply()); 
+        
+        return bytes(baseURI).length > 0 ? string(abi.encodePacked(baseURI, 
+            metadataId.toString()
+        )) : "";
+    }
+
+    
+    /** 
+        @dev Override the function to provide the maxium supply of the collection. It should be a constant value.
+
+        see also: https://docs.chain.link/docs/vrf-contracts/
+     */
+    function _maxSupply() internal pure virtual returns (uint256);
```

Comment:
Can we make this function's modifier `view` as this way we can read from the state and we can set the max supply from the constructor?

Reply:
@ivaylonikolov7 you can do something like this.

```
uint256 public maxSupply;

function _maxSupply() internal pure override returns (uint256) {
    return maxSupply;
}

Reply:
The compiler tells me that

` Function declared as pure, but this expression (potentially) reads from the environment or state and thus requires "view".` because of the way pure works I think.

Reply:
Sorry for the late reply. Does it work with
```
function _maxSupply() internal view override returns (uint256) {
```
?